apple/internal/aspects/swift_dynamic_framework_aspect.bzl

```diff
@@ -0,0 +1,144 @@
+
+load(
+    "@build_bazel_rules_swift//swift:swift.bzl",
+    "SwiftInfo",
+)
+
+SwiftDynamicFrameworkInfo = provider(
+    fields = {
+        "module_name": "The module name for the single swift_library dependency.",
+        "swiftinterfaces": """
+Dictionary of architecture to the generated swiftinterface file for that architecture.
+""",
+        "swiftdocs": """
+Dictionary of architecture to the generated swiftdoc file for that architecture.
+""",
+        "generated_header": """
+The generated Objective-C header for the single swift_library dependency.
+""",
+        "swiftmodules": """
+Dictionary of architecture to the generated swiftmodule file for that architecture.
+""",
+        "modulemap": """
+Generated modulemap for that architecture.
+""",
+    },
+    doc = """
+Provider that collects artifacts required to build a Swift-based dynamic framework.
+""",
+)
+
+def _swift_target_for_dep(dep):
+    """Returns the target for which the dependency was compiled.
+
+    This is really hacky, but there's no easy way to acquire the Apple CPU for which the target was
+    built. One option would be to let this aspect propagate transitively through deps and have
+    another provider that propagates the CPU, but the model there gets a bit more complicated to
+    follow. With this approach, we avoid propagating the aspect transitively as well.
+
+    This should be cleaned up when b/141931700 is fixed (adding support for ctx.rule.split_attr).
+    """
+    for action in dep.actions:
+        if action.mnemonic == "SwiftCompile":
+            target_found = False
+            for arg in action.argv:
+                if target_found:
+                    return arg
+                if arg == "-target":
+                    target_found = True
+    fail("error: Expected at least one Swift compilation action for target {}.".format(dep.label))
+
+def _swift_arch_for_dep(dep):
+    """Returns the architecture for which the dependency was built."""
+    target = _swift_target_for_dep(dep)
+    return target.split("-", 1)[0]
+
+def _modulemap_contents(module_name):
+    """Returns the contents for the modulemap file for the framework."""
+    return """\
+framework module {module_name} {{
+  header "{module_name}.h"
+  requires objc
+}}
+""".format(module_name = module_name)
+
+def _swift_dynamic_framework_aspect_impl(target, ctx):
+    """Aspect implementation for Swift dynamic framework support."""
+    if not hasattr(ctx.rule.attr, "deps"):
+        return []
+    
+    swiftdeps = [x for x in ctx.rule.attr.deps if SwiftInfo in x]
+    ccinfos = [x for x in ctx.rule.attr.deps if CcInfo in x]
+    # If there are no Swift dependencies, return nothing.
+    if not swiftdeps:
+        return []
+
+    # Collect all relevant artifacts for Swift dynamic framework generation.
+    module_name = None
+    generated_header = None
+    swiftdocs = {}
+    swiftmodules = {}
+    modulemap_file = None
+    for dep in swiftdeps:
```

Comment:
Should this rule only allow a single swift dep? From `ios_static_framework`:
> The Swift compiler expects a framework to contain a single Swift module, and each `swift_library` target is its own module by definition.

I think that would simplify the `module_name` and `generated_header` header checks below. Also, while thinking about this, if there were multiple swift deps, then Bazel would error out on declaring multiple files named `{module_name}_file.modulemap`.

Reply:
That is a great idea! I think you are right, it would likely error, and with the way it is currently implemented, there should only be one swift dependency so there's no issue with enforcing that.